Distribution click action mode

TaranDahl opened the pull request:
Reopen #1453 again.

### `[ ]` Distribution Mode Spread / Filter / Enable

- Now you can change the click action by using `AllowSwitchNoMoveCommand` hotkey. If the behavior to be executed by the current techno is different from the behavior displayed by the mouse, and the behavior to be executed will make the techno move near the target, the behavior will be replaced with area guard. Regardless of whether or not switch hotkey is used, default behavior can be changed through `DefaultApplyNoMoveCommand`.
- Now you can also change the click action when hold down the specific hotkey if enabled `AllowDistributionCommand`. The new behavior is like using the selected objects one by one to click on each target within the spread range.
  - The targets within the spread range will be allocated equally to the selected technos. Only when the behavior to be performed by the current techno is the same as that displayed by the mouse will it be allocated. Otherwise, it will return to the original default behavior of the game (it will not be effective for technos in the air). This will display a range ring.
  - `DefaultDistributionSpreadRange` controls the initial spread range, which is a number that's corresponding to the amount of cell radius * 512.
- `AllowDistributionCommand.SpreadMode` allows you to set spread range by hotkey. There're 4 tiers of range that can be selected by this hotkey which are identical to 0, 4, 8 and 16 cells.
  - When the range is 0, it is the original default behavior of the game.
- You can also adjust spread range by using the mouse wheel while holding down the specific hotkey if `AllowDistributionCommand.SpreadModeScroll` set to true. This allows a more precise control of spread range that each step will increase/decrease it by `DistributionSpreadScrollStep`, with 20 cells as its maximum value.
- `AllowDistributionCommand.FilterMode` allows you to set target filter by hotkey, which default to `DefaultDistributionFilterMode`.
  - When the filter is `None`, it is the default behavior of the game. If the range is not zero at this time, a green ring will be displayed. You can adjust the filter mode to:
    - `Like` - only targets with the same armor type (Completely identical `Armor`) will be selected among the targets allocated in the range. At this time, a blue ring will be displayed.
    - `Type` - only targets of the same type (like infantries, vehicles or buildings) will be selected among the targets allocated in the range. At this time, a yellow ring will be displayed.
    - `Name` - only targets of the same name (or with the same `GroupAs`) will be selected among the targets allocated in the range. At this time, a red ring will be displayed.
- `AllowDistributionCommand.AffectsAllies`, `AllowDistributionCommand.AffectsEnemies` & `AllowDistributionCommand.AffectsNeutral` allow the distribution command to work on allies (including owner), enemies or neutral target. If picking a target that's not eligible, it'll fallback to vanilla command.
- It's possible to add a button for distribution mode in the bottom bar by adding `DistributionMode` in the `ButtonList` of `AdvancedCommandBar` and `MultiplayerAdvancedCommandBar`.
  - The positions of each button are hardcoded, so it'll only decide whether enable this button or not. Distribute Mode button is now always listed after all the vanilla ones.
  - The asset of these buttons should be added in `sidec0x.mix` files which correspond to different sides, with the name `button12.shp`.
- For localization add `TXT_SWITCH_NOMOVE`, `TXT_DISTR_SPREAD`, `TXT_DISTR_FILTER`, `TXT_DISTR_HOLDDOWN`, `TXT_SWITCH_NOMOVE_DESC`, `TXT_DISTR_SPREAD_DESC`, `TXT_DISTR_FILTER_DESC`, `TXT_DISTR_HOLDDOWN_DESC`, `MSG:DistributionModeOn`, `MSG:DistributionModeOff`, `TIP:DistributionMode` into your `.csf` file.

In `rulesmd.ini`:
```ini
[GlobalControls]
AllowSwitchNoMoveCommand=false                      ; boolean
AllowDistributionCommand=false                      ; boolean
AllowDistributionCommand.SpreadMode=true            ; boolean
AllowDistributionCommand.SpreadModeScroll=true      ; boolean
AllowDistributionCommand.FilterMode=true            ; boolean
AllowDistributionCommand.AffectsAllies=true         ; boolean
AllowDistributionCommand.AffectsEnemies=true        ; boolean
AllowDistributionCommand.AffectsNeutral=true        ; boolean

[AudioVisual]
StartDistributionModeSound=                         ; sound entry
EndDistributionModeSound=                           ; sound entry
AddDistributionModeCommandSound=                    ; sound entry
```

In `ra2md.ini`:
```ini
[Phobos]
DefaultApplyNoMoveCommand=true                      ; boolean
DefaultDistributionSpreadRange=2048                 ; integer between 0 and 5120
DistributionSpreadScrollStep=16                     ; integer between 16 and 512
DefaultDistributionFilterMode=2                     ; integer, 0 - None , 1 - Like , 2 - Type , 3 - Name
```

In `uimd.ini`:
```ini
[AdvancedCommandBar]
ButtonList=[Button1],DistributionMode,[ButtonX]     ; List of button entry

[MultiplayerAdvancedCommandBar]
ButtonList=[Button1],DistributionMode,[ButtonX]     ; List of button entry
```

CrimRecya committed:
Core

CrimRecya committed:
Merge branch 'develop' into develop-Distribution

CrimRecya committed:
Update Whats-New.md

CrimRecya committed:
Draw at mouse

CrimRecya committed:
Show range ring

CrimRecya committed:
Fix a typo

CrimRecya committed:
Fix target cloaked units

CrimRecya committed:
Merge branch 'develop' into develop-Distribution

DeathFishAtEase committed:
")"

CrimRecya committed:
Add a hold down key to enable

CrimRecya committed:
Merge branch 'develop' into develop-Distribution

CrimRecya committed:
Doc

CrimRecya committed:
Fix target shrouded units

CrimRecya committed:
Fix target disguised units

CrimRecya committed:
Fix target outside units

CrimRecya committed:
Global toggle

CrimRecya committed:
Merge branch 'develop' into develop-Distribution

CrimRecya committed:
Message, sound and scroll action

CrimRecya committed:
Merge remote-tracking branch 'upstream/develop' into develop-Distribution

CrimRecya committed:
Merge remote-tracking branch 'upstream/develop' into develop-Distribution

CrimRecya committed:
Fix merge

CrimRecya committed:
Low distance first

CrimRecya committed:
...

Coronia committed:
Merge remote-tracking branch 'upstream/develop' into develop-Distribution

# Conflicts:
#	CREDITS.md
#	docs/Whats-New.md
#	src/Commands/Commands.cpp
#	src/Phobos.INI.cpp
#	src/Phobos.h

Coronia committed:
Merge branch 'develop' into develop-Distribution

CrimRecya committed:
Code style

Coronia committed:
Merge remote-tracking branch 'upstream/develop' into develop-Distribution

# Conflicts:
#	CREDITS.md
#	docs/Whats-New.md

Coronia committed:
remove unnecessary Grinder process

Coronia committed:
Merge branch 'develop' into develop-Distribution

CrimRecya committed:
Merge remote-tracking branch 'upstream/develop' into develop-Distribution

CrimRecya committed:
Refactor

CrimRecya committed:
Remove useless

CrimRecya committed:
Fix default value

CrimRecya committed:
Separate functions

CrimRecya committed:
Fix include

CrimRecya committed:
Fit with FakeOf

CrimRecya committed:
Merge branch 'develop' into develop-Distribution

CrimRecya committed:
TODO

CrimRecya committed:
Merge branch 'develop' into develop-Distribution

CrimRecya committed:
Update YRpp

CrimRecya committed:
Button on bottom

Coronia committed:
update doc

Coronia committed:
Merge branch 'develop' into develop-Distribution

Coronia committed:
revert YRpp

Coronia committed:
Merge branch 'develop' into develop-Distribution

TaranDahl committed:
Merge branch 'develop' into 分散点击模式

TaranDahl commented:
TODO:
(Summarized from https://github.com/Phobos-developers/Phobos/pull/1949#pullrequestreview-3473595716)

- [x] Create independent file for `AdvancedCommandBarButtonClass`.
- [x] Allow registering new `AdvancedCommandBarButtonClass` like `MakeCommand`.
- [ ] Rewrite the spread range code to allow exact range.
- [ ] Remove spread mode hot key.
- [ ] Press-and-drag mode

@Metadorius Please confirm the above summary, or supplement/correct the incorrect parts.

TaranDahl commented:
> there was also a comment about implementing press-and-drag mode (and not sure if the "same type" (infantry/buildings/vehicles/etc) is needed

I think there is not enough labor force to add more features.
I think the existing work is sufficient in terms of completion. We should just perfect the existing features and then merge them. 
As for more features, they should be another work, and can be implemented later, by anyone who wants them.

Metadorius commented:
> I think there is not enough labor force to add more features. I think the existing work is sufficient in terms of completion. We should just perfect the existing features and then merge them. As for more features, they should be another work, and can be implemented later, by anyone who wants them. Breaking it into two PR can also help avoid the problem of having an excessive amount of code piled up in one PR, which may lead to no one willing to review it.

This isn't a feature though? It is just a somewhat small improvement that brings it in line with how modern games do it.

- `0x6D2280` (exists in YRpp) for transforming screen coords into world coords (though Z is 0, maybe would want to account for that)
- `0x4AC4CC` is where the drag-selection band is set and the mouse hold is done
- `0x4ABCEB` is where the drag-selection release is handled
- `0xD90` in `TacticalClass` is `Rect` that contains X, Y, Width, Height

From that you could calculate 2 points, get world coords via function above, use this info to draw a corresponding circle and set the mode to such. I am not sure what is complex here. If needed I can send my decompile for this.

TaranDahl commented:
@Metadorius Is there any difference between `ShapeButtonClass` and `AdvancedCommandBarButtonClass`?

Metadorius commented:
@TaranDahl `ShapeButtonClass` is a vanilla engine class, `AdvancedCommandBarButtonClass` is something I propose to invent (perhaps a descendant from `ShapeButtonClass`) that stores the extra things that are currently stored as static arrays, for example.

TaranDahl committed:
move things to new files

TaranDahl committed:
Update Phobos.vcxproj

TaranDahl commented:
> @TaranDahl `ShapeButtonClass` is a vanilla engine class, `AdvancedCommandBarButtonClass` is something I propose to invent (perhaps a descendant from `ShapeButtonClass`) that stores the extra things that are currently stored as static arrays, for example.

Judging from the current code, there seems to be no need for it to inherit from `ShapeButtonClass`. After all, the game originally stores those attributes in a static array.
<img width="845" height="584" alt="image" src="https://github.com/user-attachments/assets/5a5d53bc-280e-4e2e-bd33-b5e85950303c" />

Metadorius commented:
> After all, the game originally stores those attributes in a static array.


Which is a bad pattern. You have class fields and methods for that.

TaranDahl commented:
Yeah I will make a new class to arrange the new buttons. But for the vanilla buttons, maybe we should just let them be?

Metadorius commented:
> Yeah I will make a new class to arrange the new buttons. But for the vanilla buttons, maybe we should just let them be?

Yeah I didn't mean we should be squeezing vanilla static array shitcode into proper classes necessarily, should be good.

TaranDahl committed:
AdvancedCommandBarButton

TaranDahl commented:
If we want to implement dragging, how should we handle the selection range?
Right now, it's several specific values.

Metadorius commented on src/Commands/DistributionMode.cpp:
```diff
@@ -0,0 +1,457 @@
+	if (!DistributionModeHoldDownCommandClass::OffMessageShowed)
+	{
+		DistributionModeHoldDownCommandClass::OffMessageShowed = true;
+		MessageListClass::Instance.PrintMessage(GeneralUtils::LoadStringUnlessMissing("MSG:DistributionModeOff", L"Distribution mode unabled."), RulesClass::Instance->MessageDelay, HouseClass::CurrentPlayer->ColorSchemeIndex, true);
```
```suggestion
		MessageListClass::Instance.PrintMessage(GeneralUtils::LoadStringUnlessMissing("MSG:DistributionModeOff", L"Distribution mode disabled."), RulesClass::Instance->MessageDelay, HouseClass::CurrentPlayer->ColorSchemeIndex, true);
```

Metadorius commented on src/Commands/DistributionMode.cpp:
```diff
@@ -0,0 +1,457 @@
+		else
+		{
+			LEA_STACK(CellStruct* const, pCell, STACK_OFFSET(0x18, 0x8));
+
+			auto invalidCell = CellStruct { -1, -1 };
+			auto pSecondCell = action == Action::Move || action == Action::PatrolWaypoint || action == Action::NoMove ? pCell : &invalidCell;
+
+			for (const auto& pSelect : ObjectClass::CurrentObjects)
+			{
+				const auto currentAction = pSelect->MouseOverCell(pCell, false, false);
+
+				DistributionModeHoldDownCommandClass::ClickedCellAction(pSelect, currentAction, pCell, pSecondCell);
+			}
+		}
+	}
+
+	Unsorted::MoveFeedback = true;
+
+	return SkipGameCode;
+}
+
+DEFINE_HOOK(0x6DBE74, TacticalClass_DrawAllRadialIndicators_DrawDistributionRange, 0x7)
+{
+	if (!DistributionModeHoldDownCommandClass::Enabled && SystemTimer::GetTime() - DistributionModeHoldDownCommandClass::ShowTime > 30)
+		return 0;
+
+	const auto spreadMode = Phobos::Config::DistributionSpreadMode;
+	const auto filterMode = Phobos::Config::DistributionFilterMode;
+
+	if (spreadMode || filterMode)
+	{
+		const auto pCell = MapClass::Instance.GetCellAt(DisplayClass::Instance.CurrentFoundation_CenterCell);
+		const auto color = (filterMode > 1)
+			? ((filterMode == 3) ? ColorStruct { 255, 0, 0 } : ColorStruct { 200, 200, 0 })
+			: ((filterMode == 1) ? ColorStruct { 0, 100, 255 } : ColorStruct { 0, 255, 50 });
+		Game::DrawRadialIndicator(false, true, pCell->GetCoords(), color, static_cast<float>(spreadMode ? (2 << spreadMode) : 0.5), false, true);
+	}
+
+	return 0;
+}
```
I think since this is so complex and spans across than a single "execute" method -- the actual distribution mode handling should not be a part of the command declaration/definition but elsewhere, and command classes should just call what they need in execute method. Same for hooks.

TaranDahl replied:
Changed

Metadorius commented on src/Commands/DistributionMode.cpp:
```diff
@@ -0,0 +1,457 @@
+				const auto pType = pTechno->GetTechnoType();
+				const int range = (2 << spreadMode);
+				const auto center = pTechno->GetCoords();
+				const auto pItems = Helpers::Alex::getCellSpreadItems(center, range);
```
You said that the code would need significant changes previously to accomodate arbitrary precise range, but I don't see what is significant here?

TaranDahl replied:
Once we used shortcut keys because there were only a few available values. If there are too many available values, then the shortcut keys will become impractical and we need to delete them. There probably won't be much actual work, but there will be changes to the existing functions.

Metadorius replied:
I think there's a misunderstanding. I think shortcut keys *could* be left in with pre-set ranges, and dragging will mean an arbitrary range is selected. Both could exist in parallel.

Although if I am being honest I personally don't think there's a need in anything else other than drag-distribution. This is how it works in many modern RTS/RTT games with advanced controls and there's no customisation, I never heard it being a problem.

TaranDahl replied:
Changed

Metadorius commented on src/Commands/DistributionMode.cpp:
```diff
@@ -0,0 +1,457 @@
+	Unsorted::MoveFeedback = false;
+}
+
+DEFINE_HOOK(0x4AE7B3, DisplayClass_ActiveClickWith_Iterate, 0x0)
```
I think this big hook should be split out into functions/methods called from a hook.

TaranDahl replied:
Changed

Metadorius reviewed (changes requested):
partial review

TaranDahl committed:
core

update

update doc

禁止分配到回收站

update

update

Update src/Commands/DistributionMode.cpp

fix

core

disable force attack for distribution mode

fix double keys breaking distribution command

fix distribution mode crash

add AffectsNeutral for distribution mode

Update Phobos.vcxproj

move things to new files

update doc

Button on bottom

TODO

Fit with FakeOf

Fix include

Separate functions

Fix default value

Remove useless

Refactor

remove unnecessary Grinder process

Code style

...

Low distance first

Fix merge

Message, sound and scroll action

Global toggle

Fix target outside units

Fix target disguised units

Fix target shrouded units

Doc

Add a hold down key to enable

Fix target cloaked units

Fix a typo

Show range ring

Draw at mouse

Core

Co-Authored-By: Kerbiter <crabiter@vivaldi.net>
Co-Authored-By: Coronia <28055583+Coronia@users.noreply.github.com>
Co-Authored-By: 绯红热茶 <169989423+CrimRecya@users.noreply.github.com>

TaranDahl committed:
update

TaranDahl committed:
update

Metadorius commented on docs/User-Interface.md:
```diff
@@ -585,6 +585,66 @@ For this command to work in multiplayer - you need to use a version of [YRpp spa
 
+### `[ ]` Distribution Mode Spread / Filter / Enable
+
+- Now you can change the click action by using `AllowSwitchNoMoveCommand` hotkey. If the behavior to be executed by the current techno is different from the behavior displayed by the mouse, and the behavior to be executed will make the techno move near the target, the behavior will be replaced with area guard. Regardless of whether or not switch hotkey is used, default behavior can be changed thr ...
```
1. I don't think this has to be in Distribution Mode docs
2. it's not clear what this does, is it a toggle for "fallback" behavior if different from the mouse one and allows disabling it? if so - maybe `UnavailableActionFallsBackToMove=action` will be more intuitive? or `UnavailableActionFallBack=none/action` (defaults to move), with customisability per type (not sure what handles actions, MissionType?) maybe useful?

Metadorius commented on docs/User-Interface.md:
```diff
@@ -585,6 +585,66 @@ For this command to work in multiplayer - you need to use a version of [YRpp spa
+### `[ ]` Distribution Mode Spread / Filter / Enable
+
+- Now you can change the click action by using `AllowSwitchNoMoveCommand` hotkey. If the behavior to be executed by the current techno is different from the behavior displayed by the mouse, and the behavior to be executed will make the techno move near the target, the behavior will be replaced with area guard. Regardless of whether or not switch hotkey is used, default behavior can be changed thr ...
+- Now you can also change the click action when hold down the specific hotkey if enabled `AllowDistributionCommand`. The new behavior is like using the selected objects one by one to click on each target within the spread range.
```
```suggestion
- You can now distribute commands across similar targets in a specified range when holding down a hotkey if `AllowDistributionCommand` is enabled. This behavior is like using the selected objects one by one to click on each target within the spread range.
```

Metadorius commented on docs/User-Interface.md:
```diff
@@ -585,6 +585,66 @@ For this command to work in multiplayer - you need to use a version of [YRpp spa
+
+- Now you can change the click action by using `AllowSwitchNoMoveCommand` hotkey. If the behavior to be executed by the current techno is different from the behavior displayed by the mouse, and the behavior to be executed will make the techno move near the target, the behavior will be replaced with area guard. Regardless of whether or not switch hotkey is used, default behavior can be changed thr ...
+- Now you can also change the click action when hold down the specific hotkey if enabled `AllowDistributionCommand`. The new behavior is like using the selected objects one by one to click on each target within the spread range.
+  - The targets within the spread range will be allocated equally to the selected technos. Only when the behavior to be performed by the current techno is the same as that displayed by the mouse will it be allocated. Otherwise, it will return to the original default behavior of the game (it will not be effective for technos in the air). This will display a range ring.
```
> Otherwise, it will return to the original default behavior of the game

do you mean the move fallback? if so - you should mention that and/or link it

> (it will not be effective for technos in the air)

hmm?

Metadorius commented on docs/User-Interface.md:
```diff
@@ -585,6 +585,66 @@ For this command to work in multiplayer - you need to use a version of [YRpp spa
+- Now you can change the click action by using `AllowSwitchNoMoveCommand` hotkey. If the behavior to be executed by the current techno is different from the behavior displayed by the mouse, and the behavior to be executed will make the techno move near the target, the behavior will be replaced with area guard. Regardless of whether or not switch hotkey is used, default behavior can be changed thr ...
+- Now you can also change the click action when hold down the specific hotkey if enabled `AllowDistributionCommand`. The new behavior is like using the selected objects one by one to click on each target within the spread range.
+  - The targets within the spread range will be allocated equally to the selected technos. Only when the behavior to be performed by the current techno is the same as that displayed by the mouse will it be allocated. Otherwise, it will return to the original default behavior of the game (it will not be effective for technos in the air). This will display a range ring.
+  - `DefaultDistributionSpreadRange` controls the initial spread range, which is a number that's corresponding to the amount of cell radius * 512.
```
we should probably move the modes description earlier
```suggestion
  - `DefaultDistributionSpreadRange` controls the initial spread range for the scrollwheel mode, which is a number that's corresponding to the amount of cell radius * 512.
```

Metadorius commented on docs/User-Interface.md:
```diff
@@ -585,6 +585,66 @@ For this command to work in multiplayer - you need to use a version of [YRpp spa
+[GlobalControls]
+AllowSwitchNoMoveCommand=false                      ; boolean
+AllowDistributionCommand=false                      ; boolean
+AllowDistributionCommand.SpreadModeHotKey=true       ; boolean
+AllowDistributionCommand.SpreadModeScroll=true      ; boolean
+AllowDistributionCommand.SpreadModeDrag=true        ; boolean
```
I don't think it should be controlled from `rules.ini`. Control schemes are not up to mod author.

The mod author should be able to enable or disable the feature (because it significantly alters the gameplay), but the players must be able to choose their preferred implementation themselves via `RA2MD.ini`.

Metadorius commented on docs/User-Interface.md:
```diff
@@ -585,6 +585,66 @@ For this command to work in multiplayer - you need to use a version of [YRpp spa
+    - `Like` - only targets with the same armor type (Completely identical `Armor`) will be selected among the targets allocated in the range. At this time, a blue ring will be displayed.
+    - `Type` - only targets of the same type (like infantries, vehicles or buildings) will be selected among the targets allocated in the range. At this time, a yellow ring will be displayed.
+    - `Name` - only targets of the same name (or with the same `GroupAs`) will be selected among the targets allocated in the range. At this time, a red ring will be displayed.
+- `AllowDistributionCommand.AffectsAllies`, `AllowDistributionCommand.AffectsEnemies` & `AllowDistributionCommand.AffectsNeutral` allow the distribution command to work on allies (including owner), enemies or neutral target. If picking a target that's not eligible, it'll fallback to vanilla command.
```
I don't think this is needed? It should derive this info from context. If I select a healer unit and distribute - it should heal any similar unit, with similar in this context being a friendly eligible target. Same for attacking, if it's hostile then attack it.

TaranDahl replied:
Too complicated to write such an algorithm

TaranDahl replied:
Moreover, I'm convinced that even if you develop such an algorithm, some people will still be dissatisfied.

Metadorius replied:
I don't see how it is complicated. You already have filtering on weapon/warhead level. Filter out those whose cursor would show the same action and be done with it. no need to overthink it.

Even then, if you think this won't be enough, those flags are the wrong customisation. You're just cutting away the feature instead of placing customisations in decisions where it's not obvious what would be better.

Metadorius commented on docs/User-Interface.md:
```diff
@@ -585,6 +585,66 @@ For this command to work in multiplayer - you need to use a version of [YRpp spa
+  - When the range is 0, it is the original default behavior of the game.
+- `AllowDistributionCommand.UseClick` controls whether distribution mode is activated by clicking on a target. When set to false, distribution mode only works via press-and-drag, and the range ring will only be shown while dragging.
+- You can also adjust spread range by using the mouse wheel while holding down the specific hotkey if `AllowDistributionCommand.SpreadModeScroll` set to true. This allows a more precise control of spread range that each step will increase/decrease it by `DistributionSpreadScrollStep`, with 20 cells as its maximum value.
+- `AllowDistributionCommand.SpreadModeDrag` allows you to adjust the spread range by pressing and dragging the mouse while holding down the specific hotkey. The drag distance from the starting point determines the spread range. This can naturally co-exist with `SpreadModeScroll`.
```
Does it issue the command on drag release? if so - it should be mentioned in documentation.

Metadorius commented on docs/User-Interface.md:
```diff
@@ -585,6 +585,66 @@ For this command to work in multiplayer - you need to use a version of [YRpp spa
+    - `Type` - only targets of the same type (like infantries, vehicles or buildings) will be selected among the targets allocated in the range. At this time, a yellow ring will be displayed.
+    - `Name` - only targets of the same name (or with the same `GroupAs`) will be selected among the targets allocated in the range. At this time, a red ring will be displayed.
+- `AllowDistributionCommand.AffectsAllies`, `AllowDistributionCommand.AffectsEnemies` & `AllowDistributionCommand.AffectsNeutral` allow the distribution command to work on allies (including owner), enemies or neutral target. If picking a target that's not eligible, it'll fallback to vanilla command.
+- It's possible to add a button for distribution mode in the bottom bar by adding `DistributionMode` in the `ButtonList` of `AdvancedCommandBar` and `MultiplayerAdvancedCommandBar`.
+  - The positions of each button are hardcoded, so it'll only decide whether enable this button or not. Distribute Mode button is now always listed after all the vanilla ones.
+  - The asset of these buttons should be added in `sidec0x.mix` files which correspond to different sides, with the name `button12.shp`.
+- For localization add `TXT_SWITCH_NOMOVE`, `TXT_DISTR_SPREAD`, `TXT_DISTR_FILTER`, `TXT_DISTR_HOLDDOWN`, `TXT_SWITCH_NOMOVE_DESC`, `TXT_DISTR_SPREAD_DESC`, `TXT_DISTR_FILTER_DESC`, `TXT_DISTR_HOLDDOWN_DESC`, `MSG:DistributionModeOn`, `MSG:DistributionModeOff`, `TIP:DistributionMode` into your `.csf` file.
```
Does this work exactly like waypoint mode hotkey? sound and a message when on/off, pressing on the bottom bar enables it on/off, and holding hotkey will also activate it (releasing deactivates).

I think this should be mentioned.

Metadorius commented on docs/User-Interface.md:
```diff
@@ -585,6 +585,66 @@ For this command to work in multiplayer - you need to use a version of [YRpp spa
+AllowDistributionCommand.AffectsNeutral=true        ; boolean
+
+[AudioVisual]
+StartDistributionModeSound=                         ; sound entry
+EndDistributionModeSound=                           ; sound entry
```
let's default this to the same sound as waypoint mode

Metadorius commented on docs/User-Interface.md:
```diff
@@ -585,6 +585,66 @@ For this command to work in multiplayer - you need to use a version of [YRpp spa
+[AudioVisual]
+StartDistributionModeSound=                         ; sound entry
+EndDistributionModeSound=                           ; sound entry
+AddDistributionModeCommandSound=                    ; sound entry
```
what's this about

Metadorius commented on docs/Whats-New.md:
```diff
@@ -991,6 +990,7 @@ HideShakeEffects=false           ; boolean
 - `<Player @ X>` can now be used as owner for pre-placed objects on skirmish and multiplayer maps (by Starkku)
 - Allow customizing charge turret delays per burst on a weapon (by Starkku)
 - Draw visual effects for airburst weapons (by CrimRecya)
+- Distribution click action mode (by CrimRecya)
```
```suggestion
- Distribution click action mode (by CrimRecya & TaranDahl)
```

Metadorius commented on docs/User-Interface.md:
```diff
@@ -585,6 +585,66 @@ For this command to work in multiplayer - you need to use a version of [YRpp spa
+- `AllowDistributionCommand.UseClick` controls whether distribution mode is activated by clicking on a target. When set to false, distribution mode only works via press-and-drag, and the range ring will only be shown while dragging.
+- You can also adjust spread range by using the mouse wheel while holding down the specific hotkey if `AllowDistributionCommand.SpreadModeScroll` set to true. This allows a more precise control of spread range that each step will increase/decrease it by `DistributionSpreadScrollStep`, with 20 cells as its maximum value.
+- `AllowDistributionCommand.SpreadModeDrag` allows you to adjust the spread range by pressing and dragging the mouse while holding down the specific hotkey. The drag distance from the starting point determines the spread range. This can naturally co-exist with `SpreadModeScroll`.
+- `AllowDistributionCommand.FilterMode` allows you to set target filter by hotkey, which default to `DefaultDistributionFilterMode`.
```
I am not sure how this tag works? at least from the documentation

TaranDahl replied:
Set it to false, and it will function exactly as you described earlier. It is fully tailored for you.

Metadorius replied:
so this is a toggle between togglable and not togglable mode? I am not sure why this needs to be customised at `rulesmd.ini` level though?

Metadorius commented on docs/User-Interface.md:
```diff
@@ -585,6 +585,66 @@ For this command to work in multiplayer - you need to use a version of [YRpp spa
 - Deselect 1 or 5 object(s) from current selected objects.
 - For localization add `TXT_DESELECT`, `TXT_DESELECT_DESC`, `TXT_DESELECT5` and `TXT_DESELECT5_DESC` into your `.csf` file.
 
+### `[ ]` Distribution Mode Spread / Filter / Enable
```
I wonder if we could assign it by default to something like C? or maybe (crazy idea) holding X (and making specifically the release of the same hotkey in a short time scattering)?

TaranDahl replied:
I believe the current architecture does not allow the same key to be used for different functions.

Metadorius reviewed (commented):
I did a doc pass, sorry, haven't got to the rest yet. treat it both as a doc review (if what I asked for is true but docs are not reflecting it yet) and design review (if what I asked about is not true yet)

TaranDahl committed:
Update docs/User-Interface.md

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

TaranDahl committed:
Update docs/Whats-New.md

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

Coronia committed:
fix compile warning and grammar

Coronia committed:
Merge branch 'develop' into 分散点击模式

# Conflicts:
#	CREDITS.md
#	docs/User-Interface.md
#	src/Ext/TechnoType/Body.h
#	src/Phobos.INI.cpp
#	src/Phobos.h
